File: FuzzySensor/src/TouchSensorSerial.java

public class TouchSensorSerial extends SerialArduino{
	//Serial Protocol
	public static final int TS_START = 255;
	public static final short TS_WRITE = 1;
	public static final short TS_READ = 0;
	public static final short TS_FORCE_BEGIN = 1;
	public static final short TS_FORCE_END = 2;
	public static final short TS_PIEZO = 3;
	public static final short TS_POT = 4;

	public static final short TS_FORCE_BEGIN_LENGTH = 4;
	public static final short TS_FORCE_END_LENGTH = 3;
	public static final short TS_PIEZO_LENGTH = 3;
	public static final short TS_PIEZOWRITE_LENGTH = 4;
	public static final short TS_POT_LENGTH = 3;
	public static final short TS_POT_WRITE_LENGTH = 4;

	public static final int TX_DELAY_TIME = 200; //Microseconds
	public static final int RX_DELAY_TIME = 10;  //Millisseconds
	
	public static int buffer[] = new int[20];
	private static int force;
	private static boolean controlTime = true;
	private static long millisInitialTime, millisFinalTime;
	
	TouchSensorSerial(){
		super();
	}
	
	private void setForce(int force) {
		this.force = force;
	}
	
	private void setMillisInitialTime(long initialTime) {
		this.millisInitialTime = initialTime;
	}
	
	private void setMillisFinalTime(long finalTime) {
		this.millisFinalTime = finalTime;
	}
	
	private void setControlTime(boolean controlTime) {
		this.controlTime = controlTime;
	}
	
	public int getForce() {
		return this.force;
	}
	
	public long getMillisInitialTime() {
		return this.millisInitialTime;
	}
	
	public long getMillisFinalTime() {
		return this.millisFinalTime;
	}
	
	public boolean getControlTime() {
		return this.controlTime;
	}
	
	private void clearBuffer(){
	  for (int  index= 0; index<buffer.length; index++)
	    buffer[index] = 0;
	}

	private int convert(int binary){
	  return binary*256;
	}
		
	@Override
	public void run() throws Exception {
		String msg = this.getInputValue();
		
		int j=0;
		for(int index = 0;index<msg.length();index++) {
			if(Character.compare(msg.charAt(index), ',') == 0) 
				j++;
			else 
				buffer[j] = 10*buffer[j] + (msg.charAt(index) - '0');
		}
			
		
		if(buffer[0] == TS_START) {
			if(buffer[1] == TS_START){
				if(buffer[2] == TS_WRITE){
			        int lengthMsg = buffer[3];
			        int type = buffer[4];
			        int checkSum;
			        switch (type){
			          case TS_FORCE_BEGIN:
			        	  int force = 0;
			        	  force = buffer[5];
			              force = force + convert(buffer[6]);
			              
			              checkSum = (~(TS_WRITE+TS_FORCE_BEGIN_LENGTH+TS_FORCE_BEGIN+(force&0xFF)+(force>>8)))&0xFF;
			              if (checkSum == buffer[7]) {
			            	  if(getControlTime()) {
			            		  setMillisInitialTime(System.currentTimeMillis());
			            		  setControlTime(false);
			            	  }
			            	  
			                setForce(force);
			                //System.out.println(getForce());
			              }
			              else
			            	throw new Exception("Falha na Mensagem");
			          
			        	  break;
			          case TS_FORCE_END:
			        	  int chr = buffer[5];
			        	  
			        	  checkSum = (~(TS_WRITE+TS_FORCE_END_LENGTH+TS_FORCE_END+chr))&0xFF;
			        	  
			        	  if (checkSum == buffer[6]) {
			            	  if(!getControlTime()) {
			            		  setMillisFinalTime(System.currentTimeMillis());
			            		  setControlTime(true);			            		 
			            	  }
			            	  setForce(0);
			              }
			              else
			            	throw new Exception("Falha na Mensagem");
			        	  break;
			        }
				}
			}
		}
		else
			throw new Exception("Erro de Comunicação");
		
		clearBuffer();
	}
	

}

File: FuzzySensor/arduinoCodes/LeituraDoSensorV2/TouchSensorSerial.cpp
#include "Arduino.h"
#include "TouchSensorSerial.h"

uint16_t TouchSensorSerial::force;
volatile unsigned long TouchSensorSerial::initialTime, TouchSensorSerial::finalTime, TouchSensorSerial::touchTime =0;
int TouchSensorSerial::sendSerial = 0;
  
/*************************************** CONSTRUCTORS ***************************************/


TouchSensorSerial::TouchSensorSerial(uint8_t IN_neg_port,uint8_t IN_pos_port){
  setIN_ports(IN_neg_port,IN_pos_port);
}

/*************************************** SETTERS ***************************************/

void TouchSensorSerial::setIN_ports(uint8_t IN_neg_port,uint8_t IN_pos_port){
  pinMode(IN_neg_port, INPUT);
  attachInterrupt(digitalPinToInterrupt(IN_neg_port), TouchSensorSerial::pulsoNegativo, RISING);
  pinMode(IN_pos_port, INPUT);
  attachInterrupt(digitalPinToInterrupt(IN_pos_port), TouchSensorSerial::pulsoPositivo, RISING);
}

/*************************************** INTERRUPTIONS ***************************************/
static void TouchSensorSerial::pulsoPositivo(){
  initialTime = millis();
  sendSerial = TS_FORCE_BEGIN;
}

static void TouchSensorSerial::pulsoNegativo(){
  finalTime = millis();
  sendSerial = TS_FORCE_END;
}

/*************************************** METHODS ***************************************/

void TouchSensorSerial::serial(int type,int ID){
  if (type == TS_FORCE_BEGIN){
    this->force = analogRead(peakPort);
    
    uint16_t *f=new uint16_t[2]{(this->force&0xff),(this->force>>8)};
    int checkSum = (~(TS_WRITE+TS_FORCE_BEGIN_LENGTH+TS_FORCE_BEGIN+ f[0]+f[1]))&0xFF;
    
    Serial.print(TS_START);
    Serial.print(",");
    Serial.print(TS_START);
    Serial.print(",");
    Serial.print(TS_WRITE);
    Serial.print(",");
    Serial.print(TS_FORCE_BEGIN_LENGTH);
    Serial.print(",");
    Serial.print(TS_FORCE_BEGIN);
    Serial.print(",");
    Serial.print(f[0]);
    Serial.print(",");
    Serial.print(f[1]);
    Serial.print(",");
    Serial.println(checkSum);
    delayMicroseconds(TX_DELAY_TIME);
  }
  else if (type == TS_FORCE_END){
    this->force = analogRead(peakPort);
   
    //unsigned long *t = new unsigned long[2]{((finalTime-initialTime)&0xff),((finalTime-initialTime)>>8)};
    int checkSum = (~(TS_WRITE+TS_FORCE_END_LENGTH+TS_FORCE_END+ 255))&0xFF;
    
    Serial.print(TS_START);
    Serial.print(",");
    Serial.print(TS_START);
    Serial.print(",");
    Serial.print(TS_WRITE);
    Serial.print(",");
    Serial.print(TS_FORCE_END_LENGTH);
    Serial.print(",");
    Serial.print(TS_FORCE_END);
    Serial.print(",");
    Serial.print(255);
    Serial.print(",");
    Serial.println(checkSum);
    delayMicroseconds(TX_DELAY_TIME);
    sendSerial = 0;
  }
  else if (type == TS_PIEZO){
    int piezoValue;
    if(ID == 'A')
      piezoValue = analogRead(piezoPort);

    uint16_t *pv = new uint16_t[2]{piezoValue&0xff,piezoValue>>8};
    int checkSum = (~(TS_WRITE+TS_PIEZOWRITE_LENGTH+TS_PIEZO+pv[0]+pv[1]))&0xFF;
    
    Serial.print(TS_START);  
    Serial.print(",");
    Serial.print(TS_START);
    Serial.print(",");
    Serial.print(TS_WRITE);
    Serial.print(",");
    Serial.print(TS_PIEZOWRITE_LENGTH);
    Serial.print(",");
    Serial.print(TS_PIEZO);
    Serial.print(",");
    Serial.print(pv[0]);
    Serial.print(",");
    Serial.print(pv[1]);
    Serial.print(",");
    Serial.println(checkSum);
    delayMicroseconds(TX_DELAY_TIME);
    sendSerial = 0;
  }
  else if (type == TS_POT){
    int potValue;
    if(ID == 'A')
      potValue = analogRead(pot_neg);
    if(ID == 'B')
      potValue = analogRead(pot_pos);

    uint16_t *pv = new uint16_t[2]{(potValue&0xff),(potValue>>8)};
    int checkSum = (~(TS_WRITE+TS_PIEZOWRITE_LENGTH+TS_PIEZO+pv[0]+pv[1]))&0xFF;
    
    Serial.print(TS_START);  
    Serial.print(",");
    Serial.print(TS_START);
    Serial.print(",");
    Serial.print(TS_WRITE);
    Serial.print(",");
    Serial.print(TS_POT_LENGTH);
    Serial.print(",");
    Serial.print(TS_PIEZO);
    Serial.print(",");
    Serial.print(pv[0]);
    Serial.print(",");
    Serial.print(pv[1]);
    Serial.print(",");
    Serial.println(checkSum);
    delayMicroseconds(TX_DELAY_TIME);
    sendSerial = 0;
  }
}

void TouchSensorSerial::verifyBuffer(int* buf){
  if(this->buf[1]== TS_START){
    if(this->buf[2] == TS_READ){
      uint8_t lengthMsg = this->buf[3];
      uint8_t type = this->buf[4];
      int checkSum;
      switch (type){
        case TS_PIEZO:
          int piezoID = 0;
          piezoID = this->buf[5];
          checkSum = (~(TS_READ+TS_PIEZO_LENGTH+TS_PIEZO+piezoID))&0xFF;
          if (checkSum == this->buf[6])
            serial(TS_PIEZO,piezoID);
          else
            Serial.println("Message Error");
          break;
          
        case TS_POT:
          int potID = 0;
          potID = this->buf[5];
          checkSum = (~(TS_READ+TS_POT_LENGTH+TS_POT+potID))&0xFF;
          if (checkSum == this->buf[6])
            serial(TS_PIEZO,potID);
          else
            Serial.println("Message Error");
          break;
      }
    }
  }
  else
    Serial.println("Error");
}

void TouchSensorSerial::clearBuffer(){
  for (uint8_t  index= 0; index<(sizeof(buf)/sizeof(int)); index++)
    buf[index] = 0;
}

int TouchSensorSerial::convert(int binary){
  if (binary == 1)
    return 256;
  else if(binary == 2)
    return 512;
  else if (binary == 3)
    return 768;
  else
    return 0;
}

void TouchSensorSerial::runSensor(){
  while (Serial.available()>0){
    this->buf[bufIndex] = Serial.read();
    /*Serial.print(bufIndex);
    Serial.print(": ");
    Serial.println(buf[bufIndex]);*/
    this->bufIndex++;
    delay(RX_DELAY_TIME);
  }

  serial(this->sendSerial,0);  

  if(this->buf[0] == TS_START && Serial.available()<=0)
    verifyBuffer(buf);
  
  bufIndex = 0;
  clearBuffer();
}

/*
void TouchSensorSerial::runSensor(){
  while (Serial.available()>0){
    this->buf[bufIndex] = Serial.read();
    Serial.print(bufIndex);
    Serial.print(": ");
    Serial.println(buf[bufIndex]);
    this->bufIndex++;
    delay(RX_DELAY_TIME);
  }
  
  serial(this->sendSerial,0);
    
  while(this->control){
    this->force = analogRead(peakPort);
    if (this->maxForce < this->force)
      this->maxForce = this->force; 
  }

  if(this->buf[0] == TS_START && Serial.available()<=0)
    verifyBuffer(buf);
  
  bufIndex = 0;
  clearBuffer();
}*/
 
